hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/source/FileIndex.java

```diff
@@ -213,24 +213,17 @@ public void setFilters(List<ResolvedExpression> filters) {
    */
   @Nullable
   private Set<String> candidateFilesInMetadataTable(FileStatus[] allFileStatus) {
-    // NOTE: Data Skipping is only effective when it references columns that are indexed w/in
-    //       the Column Stats Index (CSI). Following cases could not be effectively handled by Data Skipping:
-    //          - Expressions on top-level column's fields (ie, for ex filters like "struct.field > 0", since
-    //          CSI only contains stats for top-level columns, in this case for "struct")
-    //          - Any expression not directly referencing top-level column (for ex, sub-queries, since there's
-    //          nothing CSI in particular could be applied for)
-    if (!metadataConfig.enabled() || !dataSkippingEnabled) {
-      validateConfig();
-      return null;
+    // initialize data skipper if it's not initialized yet
+    if (this.dataPrunerOpt == null) {
+      DataPruner dataPruner = initializeDataPruner();
```

Comment:
Can we init the pruner in FileIndex constructor, make sure the `DataPruner` is serializabel.

Reply:
Of course, but it needs refactor on `HoodieTableSource`. 
Currently, the `FlieIndex` instance is created in `HoodieTableSource` constructor, but `filters` could only be got when do filter push down, the `fileIndex` is already used before filter push down. 
Just make sure if it's worth to do those refactor to init  the pruner in `FileIndex` constructor before I begin to do.